corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,652 @@
+        errors = []
+        for lt in self.location_types:
+            if lt.expand_from and lt.expand_from not in allowed_from_codes(lt.code):
+                errors.append(
+                    "'{}' can't have '{}' as 'Expand From', valid options are '{}'"
+                    .format(lt.code, lt.expand_from, allowed_from_codes)
+                )
+
+            if lt.sync_to and lt.sync_to not in allowed_to_codes(lt.code):
+                errors.append(
+                    "'{}' can't have '{}' as 'Sync To', valid options are '{}'"
+                    .format(lt.code, lt.sync_to, allowed_to_codes)
+                )
+
+        return errors
+
+    @memoized
+    def _validate_location_tree(self):
+        errors = []
+
+        def _validate_location(location):
+            loc_type = self.types_by_code.get(location.location_type)
+            if not loc_type:
+                return "Location '{}' in sheet points to a to be deleted location-type '{}'".format(
+                       location.site_code, location.location_type)
+
+            parent = self.locations_by_code.get(location.parent_code)
+            if loc_type.parent_code == ROOT_LOCATION_TYPE:
+                if parent:
+                    return "Location '{}' is a '{}' and should not have a parent".format(
+                           location.site_code, location.location_type)
+                else:
+                    return
+            else:
+                if not parent:
+                    return "Location '{}' does not have a parent set or its parent is being deleted".format(
+                           location.site_code)
+            correct_parent_type = loc_type.parent_code
+            if parent == ROOT_LOCATION_TYPE or parent.location_type != correct_parent_type:
+                return "Location '{}' is a '{}', so it should have a parent that is a '{}'".format(
+                       location.site_code, location.location_type, correct_parent_type)
+
+        for location in self.locations:
+            error = _validate_location(location)
+            if error:
+                errors.append(error)
+
+        return errors
+
+    @memoized
+    def _check_location_names(self):
+        locs_by_parent = defaultdict(list)
+        for loc in self.locations:
+            locs_by_parent[loc.parent_code].append(loc)
+        errors = []
+        for parent, siblings in locs_by_parent.items():
+            counts = Counter(l.name for l in siblings).items()
+            for name, count in counts:
+                if count > 1:
+                    errors.append(
+                        ("There are {} locations with the name '{}' under the parent '{}'"
+                         .format(count, name, parent))
+                    )
+        return errors
+
+
+def new_locations_import(domain, excel_importer):
+    try:
+        importer = NewLocationImporter.from_excel_importer(domain, excel_importer)
+    except LocationExcelSheetError, e:
```

Comment:
Actually the other reason I didn't make it so is because the types of errors are interlinked, i.e. if there is a 1st kind of error, it means we can't check for subsequent kinds of errors. There are five types of errors, only two types can be checked independent of other errors.